File: server/ai/handlers/bookingHandler.py
from .base import IntentHandler
from routes.booking_route import create

class BookingHandler(IntentHandler):
    async def handle(self, type_: str, entities: str, params: dict):
        if type_ == "reply":
            return await self.booking_text(entities, params)
        elif type_ == "ui_action":
            return await self.booking_ui(entities, params)
        elif type_ == "confirm":
            # Format các biến khi gọi create
            return "done"
            return await create(params) 
        return None

    async def booking_text(self, entities, params):
        # Tạo session (Chỉ chạy lần đầu tiên message về booking sau đó chỉ cập nhật session tới lúc confirm hoặc cancel)
        # restaurant
        restaurant = None
        if params.get("restaurant") and params["restaurant"].get("value"):
            restaurant = params["restaurant"]["value"]

        # num_people
        num_people = None
        if params.get("num_people") and params["num_people"].get("value"):
            num_people = params["num_people"]["value"]

        # contact_name
        contact_name = None
        if params.get("contact_name") and params["contact_name"].get("value"):
            contact_name = params["contact_name"]["value"]

        # contact_phone
        contact_phone = None
        if params.get("contact_phone") and params["contact_phone"].get("value"):
            contact_phone = params["contact_phone"]["value"]

        # special_request
        special_request = None
        if params.get("special_request") and params["special_request"].get("value"):
            special_request = params["special_request"]["value"]

        # promotion_code
        promotion_code = None
        if params.get("promotion_code") and params["promotion_code"].get("value"):
            promotion_code = params["promotion_code"]["value"]

        # time (time không có value, chỉ có from/to)
        time = params.get("time")  # dạng { from: null, to: null }

        # date (date không có value, chỉ có day, month, year)
        date = params.get("date")  # dạng { day: null, month: null, year: null }

        # Nếu thiếu nhiều thông tin thì cho frontend hiển thị form điền

        formatted_data = {
            "action": "create_booking",
            "updated_session": {
                "flow": "booking.create",
                "restaurant": restaurant,
                "time": time,
                "date": date,
                "num_people": num_people,
                "contact_name": contact_name,
                "contact_phone": contact_phone,
                "special_request": special_request,
                "promotion_code": promotion_code
            },
            "message": "Đây là thông tin đặt bàn của bạn. Vui lòng xác nhận để hoàn tất đặt bàn."
        }
        print("Formatted booking data:", formatted_data)
        return formatted_data

    async def booking_ui(self, entities, params):
        return {
            "type": "ui-action",
            "action": "booking",
            "entity": entities,
            "params": params
        }

File: server/ai/handlers/bookingHandler.js
import IntentHandler from "./base.js";
import Restaurant from "../../model/restaurant.js";
import Booking from "../../model/booking.js";
import BookingSlot from "../../model/bookingSlot.js";

export default class BookingHandler extends IntentHandler {
  async handle(type, entity, params) {
    if (type === "reply") {
      return await this.bookingText(params);
    } else if (type === "confirm") {
      return await this.bookingConfirm(params);
    }
    return null;
  }

  async bookingText(params) {
    // --- Lấy dữ liệu từ params ---
    const quantity = params?.quantity?.value || null;
    const booking_time = params?.booking_time || null; // { from, to }
    const booking_date = params?.booking_date || null; // ISO date string
    const restaurant = params?.restaurant?.value || null;
    const table = params?.table?.value || null;

    // Nếu thiếu nhiều thông tin thì frontend hiển thị form điền
    const formatted_data = {
      action: "create_booking",
      updated_session: {
        flow: "booking.create",
        quantity,
        booking_time,
        booking_date,
        restaurant,
        table,
      },
      message:
        "Đây là thông tin đặt bàn của bạn. Vui lòng xác nhận để tiếp tục.",
    };

    console.log("Formatted booking data:", formatted_data);
    return formatted_data;
  }
  async bookingConfirm(params) {
    try {
      const {
        restaurant,
        booking_date, // ISO string
        booking_time, // { from, to }
        table,
        quantity,
        userId,
      } = params;

      if (
        !restaurant ||
        !booking_date ||
        !booking_time ||
        !table ||
        !quantity
      ) {
        return {
          success: false,
          message: "Thiếu thông tin booking. Vui lòng điền đầy đủ.",
        };
      }

      const res = await findRestaurantByName(restaurant);
      if (!res) {
        return { success: false, message: "Không tìm thấy nhà hàng." };
      }

      const slot = await findSlotByTime(
        res._id,
        booking_time.from,
        booking_time.to
      );
      if (!slot) {
        return { success: false, message: "Không tìm thấy slot phù hợp." };
      }

      let max_table = 0;
      switch (Number(table)) {
        case 2:
          max_table = slot.max_slot_2;
          break;
        case 4:
          max_table = slot.max_slot_4;
          break;
        case 8:
          max_table = slot.max_slot_8;
          break;
        default:
          return { success: false, message: "Loại bàn không hợp lệ." };
      }

      const existBooking = await Booking.find({
        restaurant_id: res._id,
        slot_id: slot._id,
        booking_date,
        table: Number(table),
      });
      const totalBooked = existBooking.reduce((sum, b) => sum + b.quantity, 0);
      if (totalBooked + quantity > max_table) {
        const remain = max_table - totalBooked;
        return {
          success: false,
          message: `Chỉ còn ${remain} bàn loại ${table}.`,
        };
      }
      const newBooking = {
        userId,
        restaurant_id: res._id,
        slot_id: slot._id,
        booking_date,
        booking_time,
        table,
        quantity,
      };
      const booking = await Booking.create(newBooking);

      return {
        success: true,
        message: "Đặt bàn thành công!",
        booking,
      };
    } catch (err) {
      console.log(err);
      return { success: false, message: "Lỗi khi tạo booking." };
    }
  }
}

export const findRestaurantByName = async (name) => {
  if (!name) return null;

  const restaurant = await Restaurant.findOne({
    name: { $regex: new RegExp(`^${name}$`, "i") },
  }).lean();

  return restaurant;
};

export const findSlotByTime = async (restaurant_id, from, to) => {
  if (!restaurant_id || !from || !to) return null;

  const slot = await BookingSlot.findOne({
    restaurant_id,
    time: { $gte: from, $lte: to },
  }).lean();

  return slot;
};
